### PR TITLE
Ani/update training loop with lightning

### PR DESCRIPTION
**Pull request by ani0203-97**
Branched off of the Pytorch Dataloaders branch [although turns out it doesn't really depend on it]. 

This branch does the following:

- Defines a new Module that is inherited from the Pytorch Lightning module
- Creates a new script which serves as an entry point to training if we want to use Lightning 

See - https://www.notion.so/bunkerhill/Updating-Training-Loop-with-Lightning-e462486482f34443ad40e67d23734b35?pvs=4 for additional details/plan

### CODE REVIEW

**Comment by LostGeorge** [comment]
[Full review later] Should add lightning version in dependencies [https://github.com/bunkerhillhealth/nnUNet/blob/master/pyproject.toml#L32](https://github.com/bunkerhillhealth/nnUNet/blob/master/pyproject.toml#L32)

File: nnunetv2/run/run_training_lightning.py [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+import nnunetv2
+from nnunetv2.paths import nnUNet_preprocessed
+from nnunetv2.run.load_pretrained_weights import load_pretrained_weights
+from nnunetv2.training.nnUNetTrainer.nnUNetLightningModule import nnUNetLightningModule
+from nnunetv2.training.nnUNetTrainer.nnUNetTrainer import nnUNetTrainer
+from nnunetv2.utilities.dataset_name_id_conversion import maybe_convert_to_dataset_name
+from nnunetv2.utilities.find_class_by_name import recursive_find_python_class
+
+
+def find_free_network_port() -> int:
```
Comment (LostGeorge) [inline nnunetv2/run/run_training_lightning.py]: Don't think we need this anymore for DDP with lightning.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 2]
```diff
@@ -0,0 +1,869 @@
...
+        
+        # loading and saving this class for continuing from checkpoint should not happen based on pickling. This
+        # would also pickle the network etc. Bad, bad. Instead we just reinstantiate and then load the checkpoint we
+        # need. So let's save the init args
+
+        super().__init__()
+
+        self.my_init_kwargs = {}
+        for k in inspect.signature(self.__init__).parameters.keys():
+            self.my_init_kwargs[k] = locals()[k]
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:112]: [This](https://lightning.ai/docs/pytorch/2.3.3/common/lightning_module.html#id2) is lightning's built in way to save hyperparams, but not sure if it works as intended for what is passed in here. Think it's definitely better to just use the lightning functionality though.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 3]
```diff
@@ -0,0 +1,869 @@
...
+        self.is_cascaded = self.configuration_manager.previous_stage_name is not None
+        self.folder_with_segs_from_previous_stage = \
+            join(nnUNet_results, self.plans_manager.dataset_name,
+                 self.__class__.__name__ + '__' + self.plans_manager.plans_name + "__" +
+                 self.configuration_manager.previous_stage_name, 'predicted_next_stage', self.configuration_name) \
+                if self.is_cascaded else None
+
+        
+        ### Some hyperparameters for you to fiddle with
+        # TODO: Make these configurable via a separate training config file ?
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:153]: Agreed, this is important.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 4]
```diff
@@ -0,0 +1,869 @@
...
+        self.model = self.build_network_architecture(self.plans_manager, self.dataset_json, self.configuration_manager,
+                                                     self.num_input_channels, enable_deep_supervision=True)
+        
+        # compile network for free speedup
+        if self._do_i_compile():
+            self.print_to_log_file('Compiling network...')
+            self.model = torch.compile(self.model)
+
+    def _do_i_compile(self):
+        return ('nnUNet_compile' in os.environ.keys()) and (os.environ['nnUNet_compile'].lower() in ('true', '1', 't'))
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:347]: Change this to a regular hyperparam.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+        # TODO: Upgrade this to use Logging in Lightning
+        timestamp = datetime.now()
+        maybe_mkdir_p(self.output_folder)
+        self.log_file = join(self.output_folder, "training_log_%d_%d_%d_%02.0d_%02.0d_%02.0d.txt" %
+                             (timestamp.year, timestamp.month, timestamp.day, timestamp.hour, timestamp.minute,
+                              timestamp.second))
+        self.nnUNet_logger = nnUNetLogger()
+
+        ### initializing stuff for remembering things and such
+        self._best_ema = None
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:187]: What does ema mean? All I interpret from ema is "exponential moving average" but I am not that familiar with the metrics.

But in terms of best metrics and such, the [ModelCheckpoint callback](https://lightning.ai/docs/pytorch/2.3.3/api/lightning.pytorch.callbacks.ModelCheckpoint.html#modelcheckpoint) + lighting logger functionality should be able to do what is being done here.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+
+        ## Env variables
+        self.allowed_num_processes = get_allowed_n_proc_DA()
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:190]: This is specific to DKFZ stuff. We don't require it anymore.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 7]
```diff
@@ -0,0 +1,869 @@
...
+
+        ### inference things
+        self.inference_allowed_mirroring_axes = None  # this variable is set in        
+
+        ### checkpoint saving stuff
+        self.save_every = 50
+        self.disable_checkpointing = False
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:197]: See previous comment on checkpoint callback.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 8]
```diff
@@ -0,0 +1,869 @@
...
+        if do_dummy_2d_data_aug:
+            initial_patch_size[0] = patch_size[0]
+
+        self.print_to_log_file(f'do_dummy_2d_data_aug: {do_dummy_2d_data_aug}')
+        self.inference_allowed_mirroring_axes = mirror_axes
+
+        return rotation_for_DA, do_dummy_2d_data_aug, initial_patch_size, mirror_axes             
+        
+    def print_to_log_file(self, *args, also_print_to_console=True, add_timestamp=True):                
+        print_log_to_file(self.log_file, *args, also_print_to_console=also_print_to_console, add_timestamp=add_timestamp)
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:265]: Logger to replace.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 9]
```diff
@@ -0,0 +1,869 @@
...
+
+        # we give each output a weight which decreases exponentially (division by 2) as the resolution decreases
+        # this gives higher resolution outputs more weight in the loss
+        weights = np.array([1 / (2 ** i) for i in range(len(deep_supervision_scales))])
+        weights[-1] = 0
+
+        # we don't use the lowest 2 outputs. Normalize weights so that they sum to 1
+        weights = weights / weights.sum()
+        # now wrap the loss
+        loss = DeepSupervisionWrapper(loss, weights)
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:392]: Can we maybe change the name of this to `MultiScaleLossWrapper`? Current name doesn't really make much sense.

Comment (LostGeorge) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:392]: ah, I see how for some reason they just name weighted multi-scale loss as deep supervision. ugh. Still think we should change the name but nbd.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 10]
```diff
@@ -0,0 +1,869 @@
...
+        return loss
+    
+    def set_deep_supervision_enabled(self, enabled: bool):
+        """
+        This function is specific for the default architecture in nnU-Net. If you change the architecture, there are
+        chances you need to change this as well!
+        """
+        self.model.decoder.deep_supervision = enabled  
+
+    def _save_debug_information(self):
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:402]: Unsure if we should keep this function or not.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 11]
```diff
@@ -0,0 +1,869 @@
...
+        # labels can either be a list of int (regular training) or a list of tuples of int (region-based training)
+        # needed for predictions. We do sigmoid in case of (overlapping) regions
+
+        self.num_input_channels = None  # -> self.initialize()
+        self.model = None  # -> self._get_network()
+        self.model = None  
+        self.optimizer = self.lr_scheduler = None  # -> self.initialize
+        # self.grad_scaler = GradScaler() if self.device.type == 'cuda' else None
+        self.loss = None  # -> self.initialize       
+        self.nnUNet_optimizer = None
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:173]: Remove self.initialize comments, since no longer used. Remove double self.model.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 12]
```diff
@@ -0,0 +1,869 @@
...
+            loss_here = np.vstack(losses_val).mean()
+        else:
+            loss_here = np.mean(outputs_collated['loss'])
+
+        global_dc_per_class = [i for i in [2 * i / (2 * i + j + k) for i, j, k in
+                                           zip(tp, fp, fn)]]
+        mean_fg_dice = np.nanmean(global_dc_per_class)
+        self.nnUNet_logger.log('mean_fg_dice', mean_fg_dice, self.manual_current_epoch)
+        self.nnUNet_logger.log('dice_per_class_or_region', global_dc_per_class, self.manual_current_epoch)
+        self.nnUNet_logger.log('val_losses', loss_here, self.manual_current_epoch)
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:834]: FTR, all the log stuff with ddp complications can be easily done with `sync_dist` in [self.log](https://lightning.ai/docs/pytorch/2.3.3/api/lightning.pytorch.core.LightningModule.html#lightning.pytorch.core.LightningModule.log)

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 13]
```diff
@@ -0,0 +1,869 @@
...
+        self.nnUNet_optimizer = torch.optim.SGD(self.model.parameters(), self.initial_lr, weight_decay=self.weight_decay,
+                            momentum=0.99, nesterov=True)
+        self.nnUNet_lr_scheduler = PolyLRScheduler(self.nnUNet_optimizer, self.initial_lr, self.num_epochs)
+
+        self.nnUNet_optimizer.load_state_dict(checkpoint['optimizer_state'])
+
+        # Not sure about this part - I think lightning handles this internally - Leaving here for comments, will remove after
+        # if self.grad_scaler is not None:
+        #     if checkpoint['grad_scaler_state'] is not None:
+        #         self.grad_scaler.load_state_dict(checkpoint['grad_scaler_state'])
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py]: Yes, we don't need to do grad scaling with lightning.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 14]
```diff
@@ -0,0 +1,869 @@
...
+                                        momentum=0.99, nesterov=True)
+            lr_scheduler = PolyLRScheduler(optimizer, self.initial_lr, self.num_epochs)
+        else:
+            optimizer = self.nnUNet_optimizer
+            lr_scheduler = self.nnUNet_lr_scheduler
+        
+        return {"optimizer": optimizer, "lr_scheduler": lr_scheduler}
+    
+    def on_train_start(self):
+        # All of the relevant logic is already in `setup`
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:668]: Fyi, setup, prepare_data, the splitting functionality, dataloaders, and all the transform things are generally recommended (but not required necessarily) to go into a separate `LightningDataModule`. So `on_train_start` pertains to trainer + model specific things. Just wanted to comment that the hooks do have separate intentions.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py [hunk 15]
```diff
@@ -0,0 +1,869 @@
...
+        l = self.loss(output, target)
+        self.train_outputs.append({'loss': l.detach().cpu().numpy()})
+
+        return l
+
+    def on_train_epoch_end(self):
+        outputs = collate_outputs(self.train_outputs)
+
+        # TO gather and log losses in the same way as nnUNet
+        #TODO: Ideally we should not have to do this at all !!! even for logging .....
```
Comment (LostGeorge) [inline nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:721]: See other sync dist comment.

File: nnunetv2/utilities/print_to_log_file.py
Comment (LostGeorge) [inline nnunetv2/utilities/print_to_log_file.py:1]: We probably don't need this file if we use lightning loggers properly.

File: nnunetv2/training/loss/compound_losses.py [hunk 16]
```diff
@@ -55,6 +59,57 @@ def forward(self, net_output: torch.Tensor, target: torch.Tensor):
 
         result = self.weight_ce * ce_loss + self.weight_dice * dc_loss
         return result
+    
+class DC_and_CE_loss_noDDP(nn.Module):
```
Comment (LostGeorge) [inline nnunetv2/training/loss/compound_losses.py:63]: How do these noDDP losses differ? The loss functions shouldn't depend on DDP or not.

**Review by LostGeorge — COMMENTED** [review]
Good work!

Lmk there's any part I can help with.

Not sure the scope of what this PR should cover, but as I mentioned in one of the previous comments, it's generally recommended to move the data loading related things to a separate LightningDataModule. The transforms logic could even be separated out from that as well.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 2 above)
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:112]: I agree that we should use this, but for the scope of this PR I want to keep functionality as similar to nnUNet as possible. We can swap things out one by one after.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 5 above)
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:187]: you're right it is exponential moving average. 
The idea here is that it tracks an EMA of the pseudo dice, and gets the best checkpoint based on this number. 

Also on the checkpoint I agree, but want to keep it out of scope for this PR

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 6 above)
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:190]: We need the number of allowed processes so we can provide it to the new Pytorch dataloader as well though.

Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:190]: leaving it in for now, we can think of a different way to specify this number

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 9 above)
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:392]: Yeah I think it references a paper where it was introduced.

File: nnunetv2/utilities/print_to_log_file.py
Comment (ani0203-97) [reply nnunetv2/utilities/print_to_log_file.py:1]: same as above, will deprecate it after

File: nnunetv2/training/loss/compound_losses.py (hunk 16 above)
Comment (ani0203-97) [reply nnunetv2/training/loss/compound_losses.py:63]: The way current losses are written, they have the DDP logic explicitly written within them with conditionals, ie. if DDP, it computes loss a certain way. We don't need that now because Lightning will handle

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 14 above)
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:668]: Ok noted. I think in an immediate next PR we should plan to break out the Data loading related functionality into a `LightningDataModule` . My understanding was having a `LightningDataModule` is not necessary, so I avoided it to keep things as simple for the first step.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 12 above)
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:834]: We should have another PR where we swap out the logger.I think we should utilize this there. It is seeming a bit complicated for me to use it, without using the rest of the logging functionality.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 10 above)
Comment (ani0203-97) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:402]: its useful in that it outputs a file with some key training information. As of now, I do sometimes check in on the file to make sure the training run is what I expect it to be.

File: nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py (hunk 14 above)
Comment (LostGeorge) [reply nnunetv2/training/nnUNetTrainer/nnUNetLightningModule.py:668]: Yeah it's not necessary---99% of things in the LightningDataModule is implemented in LightningModule, but if we ever want to process datasets different ways and be able to swap things, the separate data module is better.

Agreed on the PR for data loading.
